types/underscore/index.d.ts

```diff
@@ -85,8 +85,11 @@ declare module _ {
 
     type EnumerableKey = string | number;
 
-    type CollectionKey<V> = V extends List<any> ? number
+    type CollectionKey<V> =
+        V extends never ? any
```

Comment:
This line is interesting. What is it for?

Reply:
This is a [trick I've been using](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45994/files#r452685140) to figure out if I have an `any` and is how I end up with [this result](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46634#discussion_r467527523). Regrettably `any` extends `List`, `Dictionary`, and `undefined`, so unless I do this I unfortunately incorrectly end up with a key type of `string | number | undefined` for an `any` collection.

I kind of wish there was an `is` operator for exact matching of types instead of just `extends` because it's hard to tell when one has a few of the built-in types and it's occasionally useful to be able to check that.

Reply:
Actually though...why am I making the result `any` here when I could make it `string | number`? Silly me is silly 😅

Reply:
Oh, darn, silly me is sadly not silly 😓. `any extends x` seems to get evaluated like a type union of every type (or maybe just the basic ones) which gets iterated over and unioned to get a result, so I end up with `string | number | undefined` unless I stick an `any` in there to eat all the results from all the other branches.

Reply:
An example of someone else noticing that the behavior where plugging `any` into a conditional type results in a union of all results across all branches of the condition is https://github.com/microsoft/TypeScript/issues/23838#issue-319720914, specifically the observation that for any type `A`,
> `any extends A ? true : false = true | false`

Reply:
Oh dear. Okay.